chapters/chapter_7/section_5.md

```diff
@@ -0,0 +1,164 @@
+
+### Example Config
+```
+source s_local {
+    file("/tmp/input.log");
+};
+
+destination d_local {
+    file("/tmp/output.log" template("$(hex ${PID}): ${MESSAGE}\n"));
+};
+
+log {
+    source(s_local);
+    destination(d_local);
+};
+```
+
+### `radix-funcs.h`
+
+```
+#ifndef RADIX_FUNCS_H
+#define RADIX_FUNCS_H
+
+#include "template/simple-function.h"
+```
+
+The inner workings of `TEMPLATE_FUNCTION_SIMPLE` are a bit much to go into but essentially they create a "construct" function, which creates and returns a `LogTemplateFunction` based on the arguments passed in. So, we need to call these macro functions, which declare the construct functions, in order to have them available to make into `Plugin` objects in the `*-plugin.c` file.
+```
+TEMPLATE_FUNCTION_PROTOTYPE(tf_radix_dec);
+TEMPLATE_FUNCTION_PROTOTYPE(tf_radix_hex);
+TEMPLATE_FUNCTION_PROTOTYPE(tf_radix_oct);
+
+#endif
+```
+
+### `radix-funcs.c`
+```
+#include <math.h>
+
+#define MAX_DIGITS 100
+
+/* Function to check the number of arguments passed in */
+static gboolean
+_check_argc(gint argc, const gchar *tf_name)
+{
+  if (argc == 0)
+    {
+      return FALSE;
+    }
+  else if (argc > 1)
+    {
+      msg_error("parsing failed: too many arguments", evt_tag_str("tf_name", tf_name));
+      return FALSE;
+    }
+  return TRUE;
+}
+
+/* Function to check the result of strtol for errors */
+static gboolean
+_check_strtol_result(gchar *endptr, const gchar *tf_name)
+{
+  if (*endptr != '\0' || errno == EINVAL || errno == ERANGE)
+    {
+      msg_error("conversion failed: invalid number", evt_tag_str("tf_name", tf_name));
+      return FALSE;
+    }
+  return TRUE;
+}
+```
+
+For this guide we will only look at implementing the template function for decimal conversion. Implementing the others is just a matter of switching a few words and is in fact done with a macro in the actual code for this plugin.
+
+`TEMPLATE_FUNCTION_SIMPLE` takes in one argument, which is a function with a specific prototype. So we just need to implement this function and pass it into the macro function.
+
+The parameters of the function are as follows:
+
+1. The current log message being parsed, which may or may not be used depending on the template function.
+2. The number of arguments passed in (not including the keyword for the template function itself).
+3. The arguments passed in (not including the keyword for the template function itself).
+4. The `GString` to which the result is appended.
+
+```
+static void
+tf_radix_dec(LogMessage *msg, gint argc, GString *argv[], GString *result)
+  {
+    const gchar *tf_name = "($dec)";
+    if (!_check_argc(argc, tf_name))
+      return;
+    gchar *endptr = g_malloc(argv[0]->len);
+    errno = 0;
+    glong original = strtol(argv[0]->str, NULL, 0);
+    if (!_check_strtol_result(endptr, tf_name))
+      return;
+    g_string_append_printf(result, "%ld", original);
```

Comment:
`endptr` needs to be freed: please add `g_free(endptr)` to the end of the function

Reply:
For this part, I realised that 1. I must have messed up updating the gitbook to the code at some point, because the call to strtol is just passing in NULL instead of `endptr`, and 2. `endptr` wasn't actually supposed to be allocated memory in the first place, since strtol overwrites the pointer itself.